From 91e7e5acd37bed62e0432a491748dfe73c77ae58 Mon Sep 17 00:00:00 2001
From: "dtucker@openbsd.org" <dtucker@openbsd.org>
Date: Mon, 5 Oct 2026 06:03:40 +0000
Subject: [PATCH] upstream: Further restrict the characters allowed in a
 command-line

supplied user name, disallowing '$' and '\'. Reported by SecBuddyF KeenLab
Tencent (CodeBuddy Security).

OpenBSD-Commit-ID: 13671c178f492af63b5c1296f284818c7809ed9a

[sshedi: port to v8.9p1]
Upstream: https://github.com/openssh/openssh-portable/commit/0c4c9a7b320bd7ff61c4a2dc345cb2e27e9663b2
Signed-off-by: Shreenidhi Shedi <shreenidhi.shedi@broadcom.com>
---
 ssh.c | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/ssh.c b/ssh.c
index 6878543..73ca9c4 100644
--- a/ssh.c
+++ b/ssh.c
@@ -617,14 +617,11 @@ valid_ruser(const char *s)
 	for (i = 0; s[i] != 0; i++) {
 		if (iscntrl((u_char)s[i]))
 			return 0;
-		if (strchr("'`\";&<>|(){}", s[i]) != NULL)
+		if (strchr("'`\";&<>|(){}$\\", s[i]) != NULL)
 			return 0;
 		/* Disallow '-' after whitespace */
 		if (isspace((u_char)s[i]) && s[i + 1] == '-')
 			return 0;
-		/* Disallow \ in last position */
-		if (s[i] == '\\' && s[i + 1] == '\0')
-			return 0;
 	}
 	return 1;
 }
-- 
2.53.0

